File: src/test/java/com/fwdekker/randomness/word/DictionaryTest.java
package com.fwdekker.randomness.word;

import nl.jqno.equalsverifier.EqualsVerifier;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;


/**
 * Unit tests for {@link Dictionary}.
 */
final class DictionaryTest {
    private Dictionary dictionary;


    @Test
    void testEmptyDictionary() {
        assertThatThrownBy(() -> Dictionary.BundledDictionary.get("dictionaries/empty.dic"))
            .isInstanceOf(IllegalArgumentException.class)
            .hasMessage("Dictionary must be non-empty.")
            .hasNoCause();
    }


    @Test
    void testGetPath() {
        useDictionary("simple");

        assertThat(dictionary.getUid())
            .isEqualTo("dictionaries/simple.dic");
    }

    @Test
    void testGetWords() {
        useDictionary("simple");

        assertThat(dictionary.getWords())
            .containsExactlyInAnyOrder("a", "the", "dog", "woof", "cat", "meow");
    }


    @Test
    void testGetWordsWithLengthInRangeInvertedRange() {
        useDictionary("simple");

        assertThat(dictionary.getWordsWithLengthInRange(696, 54))
            .isEmpty();
    }

    @Test
    void testGetWordsWithLengthInRangeNegativeLength() {
        useDictionary("simple");

        assertThat(dictionary.getWordsWithLengthInRange(-4, -2))
            .isEmpty();
    }

    @Test
    void testGetWordsWithLengthInRangeEmptyRange() {
        useDictionary("simple");

        assertThat(dictionary.getWordsWithLengthInRange(0, 0))
            .isEmpty();
    }

    @Test
    void testGetWordsWithLengthInRangeOvershortWord() {
        useDictionary("varied");

        assertThat(dictionary.getWordsWithLengthInRange(0, 1))
            .isEmpty();
    }

    @Test
    void testGetWordsWithLengthInRangeShortWord() {
        useDictionary("simple");

        assertThat(dictionary.getWordsWithLengthInRange(1, 1))
            .containsExactly("a");
    }

    @Test
    void testGetWordsWithLengthInRangeLongWord() {
        useDictionary("varied");

        assertThat(dictionary.getWordsWithLengthInRange(40, 50))
            .containsExactly("pneumonoultramicroscopicsilicovolcanoconiosis");
    }

    @Test
    void testGetWordsWithLengthInRangeOverlongWord() {
        useDictionary("simple");

        assertThat(dictionary.getWordsWithLengthInRange(1000, 1001))
            .isEmpty();
    }


    @Test
    void testGetShortestWordSimple() {
        useDictionary("simple");

        assertThat(dictionary.getShortestWord())
            .hasSize(1);
    }

    @Test
    void testGetShortestWordVaried() {
        useDictionary("varied");

        assertThat(dictionary.getShortestWord())
            .hasSize(4);
    }

    @Test
    void testGetLongestWordSimple() {
        useDictionary("simple");

        assertThat(dictionary.getLongestWord())
            .hasSize(4);
    }

    @Test
    void testGetLongestWordVaried() {
        useDictionary("varied");

        assertThat(dictionary.getLongestWord())
            .hasSize(45);
    }


    @Test
    void testCombine() {
        final Dictionary combined = Dictionary.combine(Arrays.asList(
            Dictionary.BundledDictionary.get("dictionaries/simple.dic"),
            Dictionary.BundledDictionary.get("dictionaries/varied.dic")));

        assertThat(combined.getWords())
            .containsExactlyInAnyOrder("a", "the", "dog", "woof", "cat", "meow", "simplicity", "bend",
                "consideration", "pneumonoultramicroscopicsilicovolcanoconiosis");
    }

    @Test
    void testCombineDuplicates() {
        final Dictionary combined = Dictionary.combine(Arrays.asList(
            Dictionary.BundledDictionary.get("dictionaries/simple.dic"),
            Dictionary.BundledDictionary.get("dictionaries/simple.dic")));

        assertThat(combined.getWords())
            .containsExactlyInAnyOrder("a", "the", "dog", "woof", "cat", "meow");
    }


    @Test
    void testEqualsContract() {
        EqualsVerifier.forClass(Dictionary.class)
            .usingGetClass()
            .withIgnoredFields("name", "words")
            .verify();
    }


    /**
     * Selects the {@link Dictionary} to use for the current test.
     *
     * @param dictionaryName the filename of the dictionary to use for the current test
     */
    private void useDictionary(final String dictionaryName) {
        dictionary = Dictionary.BundledDictionary.get("dictionaries/" + dictionaryName + ".dic");
    }
}

File: src/test/kotlin/com/fwdekker/randomness/word/DictionaryTest.kt
package com.fwdekker.randomness.word

import com.fwdekker.randomness.TempFileHelper
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.fail
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe
import java.util.Locale


/**
 * Returns `true` if the current operating system is Windows.
 *
 * @return `true` if the current operating system is Windows
 */
private fun isWindows() = System.getProperty("os.name").lowercase(Locale.getDefault()).contains("win")


/**
 * Unit tests for [Dictionaries][Dictionary].
 */
object DictionaryTest : Spek({
    beforeEachTest {
        BundledDictionary.clearCache()
        UserDictionary.clearCache()
    }


    describe("BundledDictionary") {
        describe("words") {
            it("fails if the file does not exist") {
                val dictionary = BundledDictionary("does_not_exist.dic")

                assertThatThrownBy { dictionary.words }
                    .isInstanceOf(InvalidDictionaryException::class.java)
                    .hasMessage("File not found.")
            }

            it("is an empty list if the dictionary is empty") {
                val dictionary = BundledDictionary("dictionaries/empty.dic")

                assertThat(dictionary.words).isEmpty()
            }

            it("contains the words of a non-empty dictionary file") {
                val dictionary = BundledDictionary("dictionaries/simple.dic")

                assertThat(dictionary.words).containsExactlyInAnyOrder("a", "the", "dog", "woof", "cat", "meow")
            }

            it("does not contain duplicate words") {
                val dictionary = BundledDictionary("dictionaries/duplicates.dic")

                assertThat(dictionary.words).containsExactlyInAnyOrder("dog", "woof", "cat", "meow")
            }

            it("ignores empty lines") {
                val dictionary = BundledDictionary("dictionaries/empty-lines.dic")

                assertThat(dictionary.words).containsExactlyInAnyOrder("woof", "meow")
            }

            it("ignores commented lines") {
                val dictionary = BundledDictionary("dictionaries/comments.dic")

                assertThat(dictionary.words).containsExactlyInAnyOrder("cat", "mouse", "tree")
            }
        }

        describe("toString") {
            it("returns a human-readable string of the dictionary's filename") {
                val dictionary = BundledDictionary("dictionaries/simple.dic")

                assertThat(dictionary.toString()).isEqualTo("dictionaries/simple.dic")
            }

            it("works even if the file does not exist") {
                val dictionary = BundledDictionary("does_not_exist.dic")

                assertThat(dictionary.toString()).isEqualTo("does_not_exist.dic")
            }
        }

        describe("equals + hash code") {
            it("equals itself") {
                val dictionary = BundledDictionary("dictionary.dic")

                assertThat(dictionary).isEqualTo(dictionary)
                assertThat(dictionary.hashCode()).isEqualTo(dictionary.hashCode())
            }

            it("equals a dictionary with the same filename") {
                val dictionary1 = BundledDictionary("dictionary.dic")
                val dictionary2 = BundledDictionary("dictionary.dic")

                assertThat(dictionary1).isEqualTo(dictionary2)
                assertThat(dictionary1.hashCode()).isEqualTo(dictionary2.hashCode())
            }

            it("does not equal a bundled dictionary with a different filename") {
                val dictionary1 = BundledDictionary("dictionary1.dic")
                val dictionary2 = BundledDictionary("dictionary2.dic")

                assertThat(dictionary1).isNotEqualTo(dictionary2)
            }

            it("does not equal a user dictionary with the same filename") {
                val dictionary1 = BundledDictionary("dictionary.dic")
                val dictionary2 = UserDictionary("dictionary.dic")

                assertThat(dictionary1).isNotEqualTo(dictionary2)
            }

            it("does not equal a different object") {
                val dictionary = BundledDictionary("dictionary.dic")
                val other = Any()

                assertThat(dictionary).isNotEqualTo(other)
            }
        }
    }

    describe("UserDictionary") {
        val tempFileHelper = TempFileHelper()


        afterGroup {
            tempFileHelper.cleanUp()
        }


        describe("words") {
            describe("validation") {
                it("fails if the file does not exist") {
                    val dictionary = UserDictionary("does_not_exist.dic")

                    assertThatThrownBy { dictionary.words }
                        .isInstanceOf(InvalidDictionaryException::class.java)
                        .hasMessage("File not found.")
                }

                it("fails if the file is deleted after construction of the dictionary") {
                    val file = tempFileHelper.createFile("blizzard\nflames\ninvest", ".dic")
                    val dictionary = UserDictionary(file.absolutePath)

                    if (!file.delete())
                        fail("Failed to delete file as part of test.")

                    assertThatThrownBy { dictionary.words }
                        .isInstanceOf(InvalidDictionaryException::class.java)
                        .hasMessage("File not found.")
                }

                it("fails if the file exists but cannot be accessed") {
                    if (isWindows()) return@it // setReadable does not work in Windows

                    val file = tempFileHelper.createFile("ladder\nkempt\npork", ".dic")
                        .also { it.setReadable(false) }
                    val dictionary = UserDictionary(file.absolutePath)

                    assertThatThrownBy { dictionary.words }
                        .isInstanceOf(InvalidDictionaryException::class.java)
                        .hasMessage("File unreadable.")
                }
            }

            describe("parsing") {
                it("is an empty list if the dictionary is empty") {
                    val file = tempFileHelper.createFile("", ".dic")
                    val dictionary = UserDictionary(file.absolutePath)

                    assertThat(dictionary.words).isEmpty()
                }

                it("contains the words of a non-empty dictionary file") {
                    val file = tempFileHelper.createFile("batmen\njollity\nbolts", ".dic")
                    val dictionary = UserDictionary(file.absolutePath)

                    assertThat(dictionary.words).containsExactlyInAnyOrder("batmen", "jollity", "bolts")
                }

                it("does not contain duplicate words") {
                    val file = tempFileHelper.createFile("dolphins\nmappings\ndolphins\nflat", ".dic")
                    val dictionary = UserDictionary(file.absolutePath)

                    assertThat(dictionary.words).containsExactlyInAnyOrder("dolphins", "mappings", "flat")
                }

                it("ignores empty lines") {
                    val file = tempFileHelper.createFile("\n\nwoof\nmeow\n\n\n", ".dic")
                    val dictionary = UserDictionary(file.absolutePath)

                    assertThat(dictionary.words).containsExactlyInAnyOrder("woof", "meow")
                }

                it("ignores commented lines") {
                    val file = tempFileHelper.createFile("# Comment\nflaming\nlove\n# Destiny\n", ".dic")
                    val dictionary = UserDictionary(file.absolutePath)

                    assertThat(dictionary.words).containsExactlyInAnyOrder("flaming", "love")
                }
            }

            describe("caching") {
                it("returns different words after the filename has been changed") {
                    val file1 = tempFileHelper.createFile("grass\nthank", ".dic")
                    val file2 = tempFileHelper.createFile("serve\nleaf", ".dic")
                    val dictionary = UserDictionary(file1.absolutePath)

                    dictionary.words
                    dictionary.filename = file2.absolutePath

                    assertThat(dictionary.words).containsExactly("serve", "leaf")
                }

                it("returns the same words even if the file is changed") {
                    val file = tempFileHelper.createFile("nest\nthose", ".dic")
                    val dictionary = UserDictionary(file.absolutePath)

                    dictionary.words
                    file.writeText("worth\nearn")

                    assertThat(dictionary.words).containsExactly("nest", "those")
                }

                it("returns the same words as another instance even if the file is changed") {
                    val file = tempFileHelper.createFile("annoy\nregular", ".dic")
                    val dictionary1 = UserDictionary(file.absolutePath)
                    val dictionary2 = UserDictionary(file.absolutePath)

                    dictionary1.words
                    file.writeText("furnish\nthan")

                    assertThat(dictionary2.words).containsExactly("annoy", "regular")
                }

                it("returns the updated words from the file after the cache is cleared") {
                    val file = tempFileHelper.createFile("seem\ngod", ".dic")
                    val dictionary = UserDictionary(file.absolutePath)

                    dictionary.words
                    file.writeText("safe\nonly")
                    UserDictionary.clearCache()

                    assertThat(dictionary.words).containsExactly("safe", "only")
                }
            }
        }

        describe("toString") {
            it("returns a human-readable string of the dictionary's filename") {
                val file = tempFileHelper.createFile("mode\ndepeche", ".dic")
                val dictionary = UserDictionary(file.absolutePath)

                assertThat(dictionary.toString()).endsWith(".dic")
            }

            it("works even if the file does not exist") {
                val dictionary = UserDictionary("does_not_exist.dic")

                assertThat(dictionary.toString()).isEqualTo("does_not_exist.dic")
            }
        }

        describe("equals + hash code") {
            it("equals itself") {
                val dictionary = UserDictionary("dictionary.dic")

                assertThat(dictionary).isEqualTo(dictionary)
                assertThat(dictionary.hashCode()).isEqualTo(dictionary.hashCode())
            }

            it("equals a dictionary with the same filename") {
                val dictionary1 = UserDictionary("dictionary.dic")
                val dictionary2 = UserDictionary("dictionary.dic")

                assertThat(dictionary1).isEqualTo(dictionary2)
                assertThat(dictionary1.hashCode()).isEqualTo(dictionary2.hashCode())
            }

            it("does not equal a user dictionary with a different filename") {
                val dictionary1 = UserDictionary("dictionary1.dic")
                val dictionary2 = UserDictionary("dictionary2.dic")

                assertThat(dictionary1).isNotEqualTo(dictionary2)
            }

            it("does not equal a bundled dictionary with the same filename") {
                val dictionary1 = UserDictionary("dictionary.dic")
                val dictionary2 = BundledDictionary("dictionary.dic")

                assertThat(dictionary1).isNotEqualTo(dictionary2)
            }

            it("does not equal a different object") {
                val dictionary = UserDictionary("dictionary.dic")
                val other = Any()

                assertThat(dictionary).isNotEqualTo(other)
            }
        }
    }
})
